pkg/reconciler/apis/apibinding/apibinding_reconcile.go

```diff
@@ -237,209 +116,300 @@ func (r *workspaceAPIExportReferenceReconciler) reconcile(ctx context.Context, a
+			apiBinding,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.InternalErrorReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Error getting APIExport %s|%s: %v",
+			apiExportClusterName,
+			workspaceRef.ExportName,
+			err,
+		)
+		return err
 	}
 
 	var boundResources []apisv1alpha1.BoundAPIResource
 	needToWaitForRequeue := false
 
 	for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
-		schema, err := r.getAPIResourceSchema(apiExportClusterName, schemaName)
-		if err != nil && apierrors.IsNotFound(err) {
-			klog.Errorf("APIResourceSchema %s|%s not found", apiExportClusterName, schemaName)
-			// TODO(ncdc): not sure if we should expose this level of detail to the binding user?
+		schema, err := c.getAPIResourceSchema(apiExportClusterName, schemaName)
+		if err != nil {
+			klog.Errorf(
+				"Error binding APIBinding %s|%s, APIExport %s|%s, APIResourceSchema %s|%s: %v",
+				apiBinding.ClusterName, apiBinding.Name,
+				apiExport.ClusterName, apiExport.Name,
+				apiExport.ClusterName, schemaName,
+				err,
+			)
+
 			conditions.MarkFalse(
 				apiBinding,
 				apisv1alpha1.APIExportValid,
-				apisv1alpha1.APIResourceSchemaNotFoundReason,
+				apisv1alpha1.InternalErrorReason,
 				conditionsv1alpha1.ConditionSeverityError,
-				"Associated APIResourceSchema %s|%s not found",
-				apiExportClusterName,
-				schemaName,
+				"Invalid APIExport. Please contact the APIExport owner to resolve",
 			)
 
-			return reconcileStatusStop, nil // don't retry
-		} else if err != nil {
-			return reconcileStatusStop, err // temporary error, retry
+			if apierrors.IsNotFound(err) {
+				return nil
+			}
+
+			return err
 		}
 
 		crd, err := crdFromAPIResourceSchema(schema)
 		if err != nil {
-			// TODO(ncdc): not sure if we should expose this level of detail to the binding user?
+			klog.Errorf(
+				"Error generating CRD for APIBinding %s|%s, APIExport %s|%s, APIResourceSchema %s|%s: %v",
+				apiBinding.ClusterName, apiBinding.Name,
+				apiExport.ClusterName, apiExport.Name,
+				apiExport.ClusterName, schemaName,
+				err,
+			)
+
 			conditions.MarkFalse(
 				apiBinding,
 				apisv1alpha1.APIExportValid,
-				apisv1alpha1.InvalidSchemaReason,
+				apisv1alpha1.InternalErrorReason,
 				conditionsv1alpha1.ConditionSeverityError,
-				"error converting APIResourceSchema %s|%s to a CRD: %v",
-				schema.ClusterName,
-				schema.Name,
-				err,
+				"Invalid APIExport. Please contact the APIExport owner to resolve",
 			)
 
-			return reconcileStatusStop, nil // don't retry
+			return nil
+		}
+
+		// Check for naming conflicts
+		nameConflictChecker := &nameConflictChecker{
+			listAPIBindings:      c.listAPIBindings,
+			getAPIExport:         c.getAPIExport,
+			getAPIResourceSchema: c.getAPIResourceSchema,
+			getCRD:               c.getCRD,
+		}
+
+		if err := nameConflictChecker.checkForConflicts(crd, apiBinding); err != nil {
+			conditions.MarkFalse(
+				apiBinding,
+				apisv1alpha1.APIsBound,
```

Comment:
on rebind this is a little odd that we set `APIsBound` to false, no? The old schemas are still bound (I assume and hope), just the new binding failed. So maybe a slightly different condition name like `APIBindingSuccessful` or so would make it clear?